lib/Sema/CSGen.cpp

```diff
@@ -2710,7 +2710,24 @@ namespace {
           tupleTypeElts.push_back(TupleTypeElt(eltTy, tupleElt.getLabel()));
         }
 
-        return setType(TupleType::get(tupleTypeElts, CS.getASTContext()));
+        Type patternTy = TupleType::get(tupleTypeElts, CS.getASTContext());
+
+        // 1. Allow optional promotion only when the matching is conditional,
+        //    i.e., not in pattern binding declarations.
+        // 2. A single-element tuple can only mean that we are matching the
+        //    payload of an enum case without splatting, in which case optional
+        //    promotion is irrelevant.
+        if (!patternBinding && tupleTypeElts.size() > 1) {
+          Type tyVar = CS.createTypeVariable(CS.getConstraintLocator(locator),
+                                             TVO_CanBindToNoEscape);
+          CS.addConstraint(
+              ConstraintKind::EqualOrOptional, tyVar, patternTy,
+              locator.withPathElement(LocatorPathElt::PatternMatch(pattern)));
```

Comment:
As far as I understand this problem what we really want is not to look through optionals but disregard the structure of the tuple pattern and let context impose it instead, so how about instead of this new constraints we actually model it this way:

1. Create a "tuple" pattern type variable (just like you did);
2. For each pattern in the tuple create a member constraint to i.e. `<tuple var>[member: .<index or name>] == <element-var>`;
3. Set tuple variable as a type of the tuple pattern.

This scheme should make it possible to infer tuple type from context and look through optionals in pattern matching context while resolving individual tuple elements.

WDYT?

Reply:
cc @hamishknight @AnthonyLatsis

Reply:
Sorry for the late reply here! I'm curious what the benefit of that approach would be, as wouldn't the new tuple member constraint need to handle the same optional unwrapping that the `EqualOrOptional` constraint currently handles, plus it would need to have tuple matching logic? Or is there something I'm missing here?

Or is the aim to have the `<tuple var>` be inferred as a non-optional? I should note that the constraint which attaches the tuple var to the surrounding pattern would fail with that, as it should only allow conversions that propagate into pattern, not out of them (though I'm becoming increasingly of the opinion that we should limit those to equality constraints where we can).

Reply:
There are no new constraints and existing member already handles unwrapping of optional base type.

Reply:
Ah, I see! Sorry, I totally missed that. I guess one worry with that approach is we don't want to allow e.g `(x: Int, y: Int)` to match with `(y: Int, x: Int)`, or even worse some nominal type `T` that just so happens to have the right members.

One other thing that we have to contend with is that we currently can use the tuple type of the pattern to introduce conversions, e.g this is currently allowed:

```swift
if case (x: 0, y: 0) = (0, 0) {}
```

Also just discovered that we currently allow duplicate labels in tuple patterns :(

```swift
if case (x: 0, x: 0) = (0, 0) {}
```

Reply:
I don’t think it should be too hard to maintain the structural matching, we can even add additional information about index/label to member constraint to make sure that the base tuple does indeed match structurally.